src/action.rs

```diff
@@ -16,14 +41,65 @@ pub enum HoldTapConfig {
     /// not used in the flow of typing, like escape for example. If
     /// you are annoyed by accidental tap, you can try this behavior.
     HoldOnOtherKeyPress,
-    /// If there is a release and a press of another key, the hold
+    /// If there is a press and release of another key, the hold
     /// action is activated.
     ///
     /// This behavior is interesting for fast typist: the different
     /// between hold and tap would more be based on the sequence of
     /// events than on timing. Be aware that doing the good succession
     /// of key might require some training.
     PermissiveHold,
+    /// A custom configuration. Allows the behavior to be controlled by a caller
+    /// supplied handler function.
+    ///
+    /// # Example:
+    /// Hold events can be prevented from triggering when pressing multiple keys
+    /// on the same side of the keyboard (but does not prevent multiple hold
+    /// events).
+    /// ```
+    /// use keyberon::action::{Action, Custom, HoldTapConfig};
+    /// use keyberon::key_code::KeyCode;
+    /// use keyberon::layout::{Stack, WaitingAction, Event};
+    ///
+    /// /// Trigger a `Tap` action on the left side of the keyboard if another
+    /// /// key on the left side of the keyboard is pressed.
+    /// fn left_mod(stack: &Stack) -> Option<WaitingAction> {
+    ///     match stack.iter().map(|s| s.event()).find(|e| e.is_press()) {
+    ///         Some(Event::Press(_, j)) if j < 6 => Some(WaitingAction::Tap),
+    ///         _ => None,
+    ///     }
+    /// }
+    ///
+    /// /// Trigger a `Tap` action on the right side of the keyboard if another
+    /// /// key on the right side of the keyboard is pressed.
+    /// fn right_mod(stack: &Stack) -> Option<WaitingAction> {
+    ///     match stack.iter().map(|s| s.event()).find(|e| e.is_press()) {
+    ///         Some(Event::Press(_, j)) if j > 5 => Some(WaitingAction::Tap),
+    ///         _ => None,
+    ///     }
+    /// }
+    ///
+    /// // Assuming a standard QWERTY layout, the left shift hold action will
+    /// // not be triggered when pressing Tab-T, CapsLock-G, nor Shift-B.
+    /// const A_SHIFT: Action = Action::HoldTap {
+    ///     timeout: 200,
+    ///     hold: &Action::KeyCode(KeyCode::LShift),
+    ///     tap: &Action::KeyCode(KeyCode::A),
+    ///     config: HoldTapConfig::Custom(Custom(&left_mod)),
+    ///     tap_hold_interval: 0,
+    /// };
+    ///
+    /// // Assuming a standard QWERTY layout, the right shift hold action will
+    /// // not be triggered when pressing Y-Pipe, H-Enter, nor N-Shift.
+    /// const SEMI_SHIFT: Action = Action::HoldTap {
+    ///     timeout: 200,
+    ///     hold: &Action::KeyCode(KeyCode::RShift),
+    ///     tap: &Action::KeyCode(KeyCode::SColon),
+    ///     config: HoldTapConfig::Custom(Custom(&right_mod)),
+    ///     tap_hold_interval: 0,
+    /// };
+    /// ```
+    Custom(CustomHandler),
```

Comment:
```suggestion
    Custom(fn(StackedIter) -> Option<WaitingAction>),
```

Reply:
I originally attempted this, but it's not possible get rid of the `CustomerHandler` type due to trait errors on fns:
```
error: implementation of `Debug` is not general enough
  --> src/action.rs:99:12
   |
29 | #[derive(Debug, Clone, Copy, Eq, PartialEq)]
   |          ----- in this derive macro expansion
...
99 |     Custom(for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `Debug` is not general enough
   |
   = note: `Debug` would have to be implemented for the type `for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>`
   = note: ...but `Debug` is actually implemented for the type `fn(StackedIter<'0>) -> Option<WaitingAction>`, for some specific lifetime `'0`
   = note: this error originates in the derive macro `Debug` (in Nightly builds, run with -Z macro-backtrace for more info)

error: implementation of `core::cmp::Eq` is not general enough
  --> src/action.rs:99:12
   |
29 | #[derive(Debug, Clone, Copy, Eq, PartialEq)]
   |                              -- in this derive macro expansion
...
99 |     Custom(for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `core::cmp::Eq` is not general enough
   |
   = note: `core::cmp::Eq` would have to be implemented for the type `for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>`
   = note: ...but `core::cmp::Eq` is actually implemented for the type `fn(StackedIter<'0>) -> Option<WaitingAction>`, for some specific lifetime `'0`
   = note: this error originates in the derive macro `Eq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0369]: binary operation `==` cannot be applied to type `for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>`
  --> src/action.rs:99:12
   |
29 | #[derive(Debug, Clone, Copy, Eq, PartialEq)]
   |                                  --------- in this derive macro expansion
...
99 |     Custom(for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0369]: binary operation `!=` cannot be applied to type `for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>`
  --> src/action.rs:99:12
   |
29 | #[derive(Debug, Clone, Copy, Eq, PartialEq)]
   |                                  --------- in this derive macro expansion
...
99 |     Custom(for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Reply:
OK, that's https://users.rust-lang.org/t/why-does-not-a-function-generic-over-a-lifetime-implement-debug/17828

The "clean" solution would be to implement by hand Debug and PartialEq on Action.

Reply:
Just pushed. And right about to leave for vacation. Feel free to make any additional changes or wait for me to come back in a week. Thanks for the thorough reviews!